geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/SlotAdvisor.java

```diff
@@ -0,0 +1,210 @@
+
+  private static final Logger logger = LogService.getLogger();
+  private static final int HOSTPORT_RETRIEVAL_ATTEMPTS = 10;
+  private static final int HOSTPORT_RETRIEVAL_INTERVAL = 6_000;
+
+  /**
+   * Cache of member ids to member IP address and redis listening port
+   */
+  private final Map<DistributedMember, Pair<String, Integer>> hostPorts = new HashMap<>();
+  private final PartitionedRegion dataRegion;
+
+  SlotAdvisor(Region<RedisKey, RedisData> dataRegion) {
+    this.dataRegion = (PartitionedRegion) dataRegion;
+  }
+
+  public boolean isLocal(RedisKey key) {
+    return dataRegion.getRegionAdvisor().getBucket(key.getBucketId()).getBucketAdvisor()
+        .isPrimary();
+  }
+
+  public Pair<String, Integer> getHostAndPortForKey(RedisKey key) {
+    return getHostPort(key.getBucketId());
+  }
+
+  public Map<String, List<Integer>> getMemberBuckets() {
+    initializeBucketsIfNecessary();
+
+    Map<String, List<Integer>> memberBuckets = new HashMap<>();
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      String memberId = getOrCreateMember(bucketId).getUniqueId();
+      memberBuckets.computeIfAbsent(memberId, k -> new ArrayList<>()).add(bucketId);
+    }
+
+    return memberBuckets;
+  }
+
+  /**
+   * This returns a list of {@link MemberBucketSlot}s where each entry corresponds to a bucket. If
+   * the details for a given bucket cannot be determined, that entry will contain {@code null}.
+   */
+  public synchronized List<MemberBucketSlot> getBucketSlots() {
+    initializeBucketsIfNecessary();
+
+    List<MemberBucketSlot> memberBucketSlots = new ArrayList<>(RegionProvider.REDIS_REGION_BUCKETS);
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      Pair<String, Integer> hostPort = getHostPort(bucketId);
+      if (hostPort != null) {
+        memberBucketSlots.add(
+            new MemberBucketSlot(bucketId, hostPort.getLeft(), hostPort.getRight()));
+      }
+    }
+
+    return memberBucketSlots;
+  }
+
+  private InternalDistributedMember getOrCreateMember(int bucketId) {
+    return dataRegion.getOrCreateNodeForBucketWrite(bucketId, null);
+  }
+
+  private void initializeBucketsIfNecessary() {
+    if (dataRegion.getDataStore() != null &&
+        dataRegion.getDataStore().getAllLocalBucketIds().isEmpty()) {
+      PartitionRegionHelper.assignBucketsToPartitions(dataRegion);
+    }
+  }
+
+  /**
+   * This method will retry for {@link #HOSTPORT_RETRIEVAL_ATTEMPTS} attempts and return null if
+   * no information could be retrieved.
+   */
+  private Pair<String, Integer> getHostPort(int bucketId) {
+    Pair<String, Integer> response;
+
+    for (int i = 0; i < HOSTPORT_RETRIEVAL_ATTEMPTS; i++) {
+      response = getHostPort0(bucketId);
+      if (response != null) {
+        return response;
+      }
+
+      try {
+        Thread.sleep(HOSTPORT_RETRIEVAL_INTERVAL);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
```

Comment:
This should probably just throw interrupted exception up to the caller. Wrapping it in a runtime exception means we'll might get some weird errors during a shutdown or something I think. Other option would be to ignore it, finish the loop and then reset the interrupt bit (Call Thread.interrupt()).

Reply:
I think it's improved now. Exceptions will ultimately be handled in `ExecutionHandlerContext.exceptionCaught`.

Reply:
Is writing an error back to the client the right way to handle an interrupt though?